crates/rpc-types-engine/src/superchain.rs

```diff
@@ -0,0 +1,146 @@
+/// specifications. Components of the OP-Stack implement the subset of their respective protocol
+/// component domain, up to a given Protocol Version of the OP-Stack.
+///
+/// The Protocol Version **is NOT a hardfork identifier**, but rather indicates software-support for
+/// a well-defined set of features introduced in past and future hardforks, not the activation of
+/// said hardforks.
+///
+/// The Protocol Version is Semver-compatible. It is encoded as a single 32 bytes long
+/// protocol version. The version must be encoded as 32 bytes of DATA in JSON RPC usage.
+///
+/// See also: <https://specs.optimism.io/protocol/superchain-upgrades.html#protocol-version>
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ProtocolVersion {
+    /// Version-type 0.
+    V0(ProtocolVersionFormatV0),
+}
+
+cfg_if! {
+    if #[cfg(feature = "serde")] {
+        use serde::{de, Serialize, Serializer, Deserialize, Deserializer};
+
+        impl Serialize for ProtocolVersion {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                match self {
+                    Self::V0(value) => {
+                        // <protocol version> ::= <version-type><typed-payload>
+                        // <version-type> ::= <uint8>
+                        // <typed-payload> ::= <31 bytes>
+                        let mut bytes = [0u8; 32];
+                        bytes[0] = 0x00; // this is not necessary, but addded for clarity
+                        bytes[1..].copy_from_slice(&Into::<[u8; 31]>::into(*value));
+
+                        B256::from_slice(&bytes).serialize(serializer)
+                    }
+                }
+            }
+        }
+
+        impl<'de> Deserialize<'de> for ProtocolVersion {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                // The version must be encoded as 32 bytes of DATA in JSON RPC usage.
+                let value = B256::deserialize(deserializer)?;
+
+                // <protocol version> ::= <version-type><typed-payload>
+                // <version-type> ::= <uint8>
+                // <typed-payload> ::= <31 bytes>
+                let version_type = value[0];
+                let typed_payload = &value[1..];
+
+                match version_type {
+                    0 => Ok(Self::V0(ProtocolVersionFormatV0::try_from(typed_payload).map_err(de::Error::custom)?)),
+                    other => Err(de::Error::custom(format!("unsupported protocol version: {}", other))),
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ProtocolVersionFormatV0 {
+    pub build: B64,
+    pub major: u32,
+    pub minor: u32,
+    pub patch: u32,
+    pub pre_release: u32,
+}
+
+impl From<ProtocolVersionFormatV0> for [u8; 31] {
```

Comment:
let's impl this as function, from conversions like this are just obfuscating things.